### ISSUE TITLE
SimStore data loss after crash

### ISSUE THREAD

**Issue by Bernadette-Mohr** [openpathsampling/openpathsampling#1131]
Hi all,

My TPS simulation crashed between two save steps. As a result, the SimStore database is partially corrupted.
Is there any way to retrieve the first n entries up to this point?

```pytb
scheme.move_summary(storage.steps) 
  File "/home/bmohr/micromamba/envs/ops/lib/python3.9/site-packages/openpathsampling/high_level/move_scheme.py", line 888, in move_summary
    self._mover_acceptance.add_steps(steps)
  File "/home/bmohr/micromamba/envs/ops/lib/python3.9/site-packages/openpathsampling/high_level/move_scheme.py", line 73, in add_steps
    for step in self.progress(steps):
  File "/home/bmohr/micromamba/envs/ops/lib/python3.9/site-packages/tqdm/std.py", line 1178, in __iter__
    for obj in iterable:
  File "/home/bmohr/micromamba/envs/ops/lib/python3.9/site-packages/openpathsampling/experimental/simstore/storage.py", line 582, in __iter__
    loaded = self.storage.load(row_uuids)
  File "/home/bmohr/micromamba/envs/ops/lib/python3.9/site-packages/openpathsampling/experimental/simstore/storage.py", line 411, in load
    new_uuids = self.deserialize_uuids(ordered_uuids, uuid_to_table,
  File "/home/bmohr/micromamba/envs/ops/lib/python3.9/site-packages/openpathsampling/experimental/simstore/storage.py", line 434, in deserialize_uuids
    table = uuid_to_table[uuid]
KeyError: '142852363313043608565700386554887206642'
```

Since issues can unfortunately always happen, it would be very unfortunate if data couldn't be salvaged. 
I didn't find much documentation for the experimental storage module, and it doesn't know many of the generic SQL methods like query or limit.

I would appreciate any help!

**Comment by dwhswenson** [comment]
Just to clarify: does the error happen only when you get to the last step? That is, does it seem to do the analysis until it hits the last, corrupted, step?

If that's the case, as a first try, see if either of the following works:

```python
steps = storage.steps[:-1]
```

or

```python
nsteps = len(storage.steps)
steps = storage.steps[:nsteps]
```

Then continue by using this `steps` variable in place of `storage.steps` in your analysis. This should load up everything except that last step.

**Comment by Bernadette-Mohr** [comment]
The error occurs when it hits the first corrupted step. Unfortunately, slicing doesn't do it, that was my first attempt as well:
`Traceback (most recent call last):
  File "/home/bmohr/Data/PycharmProjects/WC2HG/recover_database.py", line 44, in <module>
    scheme.move_summary(storage.steps[0:500])  # just returns some statistics
  File "/home/bmohr/micromamba/envs/ops/lib/python3.9/site-packages/openpathsampling/experimental/simstore/storage.py", line 590, in __getitem__
    if not (-len_self <= item < len_self):
TypeError: '<=' not supported between instances of 'int' and 'slice'`

**Comment by dwhswenson** [comment]
Ah, slicing support was added in code that isn't yet in a release. If you install OPS from the current GitHub `master`, you would have slicing support. (However, it is a little risky to switch versions with existing stored data -- should be okay, but safest to stay with the version you have.)

A couple other options:

```python
steps = []
try:
    for step in storage.steps:
        steps.append(step)
except KeyError:
    pass
```

This will give you everything up to the first step that fails to load (which we assume is the final, incomplete, step).

Another approach, if that gives any problems:

```python
steps = []
for num in range(len(steps)):
    try:
        steps.append(storage.steps[num])
    except Exception as e:
        print(f"Unable to load step {num}: {e.__class__}: {str(e)}")
``` 

This should give you information about all steps that can't be loaded (especially useful if it happens that it isn't only the final step that is a problem.)

Overall theme here is to generate the list of steps that *are* good, and then all the analysis tools can take that list as input.

**Comment by Bernadette-Mohr** [comment]
Thanks, this works! 
I didn't know I could pass the steps as a list to `move_summary` since `storage.steps` seems to be an SQL table. Which is arguably not my strong suit.
Is there any way to copy the entire database file up to the corrupted part into a new file that can be used for a continuation run?

**Comment by dwhswenson** [comment]
Double checked, and a few minor suggestions for improvements on my original suggestion for copying the database:

1. Instead of `for cv in old_storage.cvs:`, do `for cv in old_storage.storable_functions:` -- this will not only copy the precalculated values of the CVs (which map a snapshot to a number), but all functions, such as the max-lambda functions used in TIS (which map a trajectory to a number).
2. The loop over steps can take some time (and there are probably some improvement we need to do to make that faster). If you'll be watching this in real time, I suggest adding a progress bar. I think OPS comes with `tqdm` installed, so change the loop over `steps` to:

```python
from tqdm.auto import tqdm
for step in tqdm(steps):
    new_storage.save(step)
```

---

tl;dr: to copy from `old_storage` to `new_storage`, after having made the `steps` variable with the valid steps as described above, do the following (testing against current `master`):

```python
for cv in old_storage.storable_functions:
    cv.preload_cache()

# this might not be required, but won't hurt
for obj in old_storage.simulation_objects:
    new_storage.save(obj)

from tqdm.auto import tqdm  # this will give a progress bar
for step in tqdm(steps):
    new_storage.save(step)

new_storage.sync_all()  # not sure if this is needed with SimStore
```

There is a possibility that something weird will happen here because of the corrupted data (I haven't tested against that), so let me know if after this you get issues (mostly likely `KeyError`s).

**Comment by Bernadette-Mohr** [comment]
It's been a while, but I still wanted to give some feedback!
Your latest suggestion works as intended!
And good call to add the progress bar. The copy process does take a while and pretty much monopolizes the RAM. I'd recommend starting the process on a Friday on a machine that won't be needed for anything else over the weekend.
